tools/resgen/src/main.cpp

```diff
@@ -0,0 +1,270 @@
+    transform(packagePrefix.begin(), packagePrefix.end(), packagePrefix.begin(), ::toupper);
+
+    const Path deployDir(g_deployDir);
+    const Path appleAsmPath(deployDir + (packageFile + ".apple.S"));
+    const Path asmPath(deployDir + (packageFile + ".S"));
+    const Path binPath(deployDir + (packageFile + ".bin"));
+    const Path headerPath(deployDir + (packageFile + ".h"));
+    const Path xxdPath(deployDir + (packageFile + ".c"));
+
+    // In the assembly language templates, replace {RESOURCES} with packagePrefix and replace
+    // {resources} with packageFile.
+    const string k1("{RESOURCES}");
+    const string k2("{resources}");
+    string aasmstr(APPLE_ASM_TEMPLATE);
+    string asmstr(ASM_TEMPLATE);
+    for (size_t pos = aasmstr.find(k1); pos != string::npos; pos = aasmstr.find(k1, pos))
+        aasmstr.replace(pos, k1.length(), packagePrefix);
+    for (size_t pos = aasmstr.find(k2); pos != string::npos; pos = aasmstr.find(k2, pos))
+        aasmstr.replace(pos, k2.length(), packageFile);
+    for (size_t pos = asmstr.find(k1); pos != string::npos; pos = asmstr.find(k1, pos))
+        asmstr.replace(pos, k1.length(), packagePrefix);
+    for (size_t pos = asmstr.find(k2); pos != string::npos; pos = asmstr.find(k2, pos))
+        asmstr.replace(pos, k2.length(), packageFile);
+
+    // Generate the Apple-friendly assembly language file.
+    ofstream appleAsmStream(appleAsmPath.getPath());
+    if (!appleAsmStream) {
+        cerr << "Unable to open " << appleAsmPath << endl;
+        exit(1);
+    }
+    appleAsmStream << aasmstr << endl;
+    appleAsmStream.close();
+
+    // Generate the non-Apple assembly language file.
+    ofstream asmStream(asmPath.getPath());
+    if (!asmStream) {
+        cerr << "Unable to open " << asmPath << endl;
+        exit(1);
+    }
+    asmStream << asmstr << endl;
+    asmStream.close();
+
+    // Open the bin file for writing.
+    ofstream binStream(binPath.getPath(), ios::binary);
+    if (!binStream) {
+        cerr << "Unable to open " << binPath << endl;
+        exit(1);
+    }
+
+    // Open the header file for writing.
+    ofstream headerStream(headerPath.getPath());
+    if (!headerStream) {
+        cerr << "Unable to open " << headerPath << endl;
+        exit(1);
+    }
+    headerStream << "#ifndef " << packagePrefix << "H_" << endl
+            << "#define " << packagePrefix << "H_" << endl << endl
+            << "extern \"C\" {" << endl
+            << "    extern const uint8_t " << packagePrefix << "PACKAGE[];" << endl
+            << "    extern const int " << packagePrefix << "PACKAGE_SIZE;" << endl
+            << "}" << endl << endl;
+
+    // Open the generated C file for writing.
+    ofstream xxdStream(xxdPath.getPath());
+    if (!xxdStream) {
+        cerr << "Unable to open " << xxdPath << endl;
+        exit(1);
+    }
+    xxdStream << "const uint8_t " << packagePrefix << "PACKAGE[] = {" << endl;
```

Comment:
Perhaps WebAssembly doesn't need this, but maybe add a `#include <stdint.h>` at the beginning of the xxdstream otherwise it doesn't compile with Clang (`uint8_t` is undefined).

Reply:
Nice catch, done.